### ISSUE TITLE
0x52 - Rollup address aliasing blocks users from repaying/topping up during sequencer downtime

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-10-real-wagmi-judging#77]
0x52

medium

# Rollup address aliasing blocks users from repaying/topping up during sequencer downtime
## Summary

Arbitrum and other popular rollups [alias addresses](https://docs.arbitrum.io/arbos/l1-to-l2-messaging#address-aliasing) when conducting L1 -> L2 transactions. These type of transactions are essential when trying to complete an L2 transaction when the sequencer is down. Due the borrower restriction on both increaseCollateralBalance and repay it is impossible to do either if the sequencer goes down. On the other hand the user can still be liquidated by anyone else, causing unfair loss to the user. 

## Vulnerability Detail

[LiquidityBorrowingManager.sol#L559-L561](https://github.com/sherlock-audit/2023-10-real-wagmi/blob/main/wagmi-leverage/contracts/LiquidityBorrowingManager.sol#L559-L561)

            (msg.sender != borrowing.borrower && collateralBalance >= 0).revertError(
                ErrLib.ErrorCode.INVALID_CALLER
            );

First we consider repay. This will revert if msg.sender is not the borrower and the collateralBalance is positive. In the event of sequencer downtime, their address will be aliased and the transaction will revert if they are still solvent.

[LiquidityBorrowingManager.sol#L371-L376](https://github.com/sherlock-audit/2023-10-real-wagmi/blob/main/wagmi-leverage/contracts/LiquidityBorrowingManager.sol#L371-L376)

    function increaseCollateralBalance(bytes32 borrowingKey, uint256 collateralAmt) external {
        BorrowingInfo storage borrowing = borrowingsInfo[borrowingKey];
        // Ensure that the borrowed position exists and the borrower is the message sender
        (borrowing.borrowedAmount == 0 || borrowing.borrower != address(msg.sender)).revertError(
            ErrLib.ErrorCode.INVALID_BORROWING_KEY
        );

Next we look at increaseCollateralBalance. This has the exact same issue, blocking the user from either topping up their position or safely exiting their position. Once their collateral is used up, they will be unfairly liquidated causing loss to the borrower.

## Impact

During sequencer downtime, borrowers will be unfairly liquidated

## Code Snippet

[LiquidityBorrowingManager.sol#L532-L674](https://github.com/sherlock-audit/2023-10-real-wagmi/blob/main/wagmi-leverage/contracts/LiquidityBorrowingManager.sol#L532-L674)

## Tool used

Manual Review

## Recommendation

When checking msg.sender in increaseCollateralBalance and repay, check both borrower and aliased borrower address. This allows borrowers to always increase their collateral or repay their loan even when the sequencer is down and their address is aliased.

**Comment by IAm0x52** [comment]
Escalate

This was incorrectly excluded. I have clearly demonstrated a scenario in which a borrower can be liquidated while being unable to repay or top up their loan. Historically this has always be judged as a valid medium issue.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> This was incorrectly excluded. I have clearly demonstrated a scenario in which a borrower can be liquidated while being unable to repay or top up their loan. Historically this has always be judged as a valid medium issue.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Czar102** [comment]
Sherlock's rules seem not to describe what should be the assumption with respect to the liveness of the network sequencer well.
> Chain re-org related issues are not considered valid.

This can make me think that liveness/consensus issues are out of scope.

Also, if the project is deployed on Arbitrum, it means that it considers the sequencer a type of admin of the network it is on (the protocol it integrates with). Hence, the sequencer is trusted and should be assumed not to misbehave (go down).

This also is a design choice to some extent. There are also some exploit scenarios connected with allowing the alias address to call the contract. For example, on both Arbitrum and L1 there may be a contract under the same address, but they may be controlled by different parties or even be deployed entirely independently (like using CREATE3 factories).

Hence, I think this should stay low/informational. The protocol team should consider adding this useful functionality, though.

I am aware of [this](https://github.com/sherlock-audit/2023-05-Index-judging/issues/321) past issue thanks to @cvetanovv, but docs, readme and general rules take precedence over past judgements.

Planning to reject the escalation and leave the issue as is.

**Comment by nevillehuang** [comment]
Imo, current sherlock docs, contest readme and general rules are still not clear enough to simply invalidate sequencer downtime issues. Not too sure sequencer downtime automatically directly translates to being related to a chain reorg. 

I think @fann95 should take a look at this, if not I will respect the final escalation outcome.

**Comment by fann95** [comment]
> Sherlock's rules seem not to describe what should be the assumption with respect to the liveness of the network sequencer well.
> 
> > Chain re-org related issues are not considered valid.
> 
> This can make me think that liveness/consensus issues are out of scope.
> 
> Also, if the project is deployed on Arbitrum, it means that it considers the sequencer a type of admin of the network it is on (the protocol it integrates with). Hence, the sequencer is trusted and should be assumed not to misbehave (go down).
> 
> This also is a design choice to some extent. There are also some exploit scenarios connected with allowing the alias address to call the contract. For example, on both Arbitrum and L1 there may be a contract under the same address, but they may be controlled by different parties or even be deployed entirely independently (like using CREATE3 factories).
> 
> Hence, I think this should stay low/informational. The protocol team should consider adding this useful functionality, though.
> 
> I am aware of [this](https://github.com/sherlock-audit/2023-05-Index-judging/issues/321) past issue thanks to @cvetanovv, but docs, readme and general rules take precedence over past judgements.
> 
> Planning to reject the escalation and leave the issue as is.

I need to think about it, I’m not yet sure about the feasibility of this functionality.
Even if I add an alias check according to [this](https://docs.arbitrum.io/sequencer#unhappyuncommon-case-sequencer-isnt-doing-its-job) information, we will be able to use it only after 24 hours of sequester downtime.

**Comment by IAm0x52** [comment]
> Sherlock's rules seem not to describe what should be the assumption with respect to the liveness of the network sequencer well.
> 
> > Chain re-org related issues are not considered valid.
> 
> This can make me think that liveness/consensus issues are out of scope.
> 

Sequencer downtime has never been considered as "liveness" before. The key difference is that if L1 like SOL or ETH go down there is no way to interact with the protocol. In the case of rollups like OP and ARB. The contracts can still be interacted with through forced inclusion. This creates an asymmetry that should be accounted for.

> Also, if the project is deployed on Arbitrum, it means that it considers the sequencer a type of admin of the network it is on (the protocol it integrates with). Hence, the sequencer is trusted and should be assumed not to misbehave (go down).
> 

The sequencer going down is not "misbehavior", it can go down for a number of reasons such as malicious targeting.

> This also is a design choice to some extent. There are also some exploit scenarios connected with allowing the alias address to call the contract. For example, on both Arbitrum and L1 there may be a contract under the same address, but they may be controlled by different parties or even be deployed entirely independently (like using CREATE3 factories).
> 

Being able to generate an aliased address would be a extremely unlikely considering the collision resistance in this scenario is 2^160. This change does not introduce any additional risk.

**Comment by detectiveking123** [comment]
I am aligned with @Czar102 on this one. I also believe the sequencer should be trusted.

**Comment by Czar102** [comment]
> The sequencer going down is not "misbehavior", it can go down for a number of reasons such as malicious targeting.

No matter what the reason is, it wouldn't execute its function. Malicious targeting also is out of scope if we consider the sequencer an external admin, which is trusted.
> Being able to generate an aliased address would be a extremely unlikely considering the collision resistance in this scenario is 2^160. This change does not introduce any additional risk.

I wasn't talking about finding a collision. You can have the same EOA address on different chains. There could be a factory deploying any contracts, with the same address on all chains (deployed by the same EOA with the same "account state nonce"). No matter if it creates contracts using CREATE or CREATE3 (like [here](https://github.com/transmissions11/solmate/blob/main/src/utils/CREATE3.sol)), there could be a cross-chain collision of the smart contract address (the data used to derive the address was the same). One would be a protocol user on Arbitrum and the second would be able to modify positions on behalf of that contract… And they may be controlled by different parties not trusting each other.

If a position can't be hurt by that action, the access may as well be open to anyone. But don't know it that's the case.

Anyway, my primary argument is that sequencer of an EVM network should be considered an external admin (as a whitelisted party in a protocol the codebase integrates with). Hence we assume it doesn't do harm (keeps the network alive).

**Comment by nevillehuang** [comment]
If the above arguments on primary issue being sequencer as a trusted external admin turns out to affect validity of issue, sherlock rules need to be updated accordingly to reflect it.

Heres a link to an ongoing discussion for sequencer downtime on the aloe contest for further consideration.

https://github.com/sherlock-audit/2023-10-aloe-judging/issues/55

**Comment by Evert0x** [comment]
The current language in the judging guidelines and previous judgments on similar issues are not sufficient to arrive at a conclusion considering the context of the protocol deployment. 

For this reason, the following adjustment had been made to the judging guidelines: https://docs.sherlock.xyz/audits/judging/judging/criteria-changelog#1.4-10-11-2023

The underlying reason for this change is that issues like this are not always valuable to the protocol team, especially if the protocol team fully trusts the context of the deployment. This also signals to end users that the protocol is written in a way where these trust assumptions exist.

Plan to reject escalation and keep issue state as is considering the protocol team put TRUSTED on the external admins question.

**Comment by Evert0x** [comment]
Result:
Invalid
Unique

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [IAm0x52](https://github.com/sherlock-audit/2023-10-real-wagmi-judging/issues/77/#issuecomment-1789399141): rejected
